### PR TITLE
Fix integer cast in omrerror.c

### PR DESCRIPTION
**Pull request by ktbriggs**
Port library method str_vprintf() requires a uintptr_t value for buffer
length and returns uintptr_t, return value was being assigned to
uint32_t value that is later assigned to a uint32_t-valued
field. 

Buffer length arg for str_vprintf() is not used when buffer pointer is
NULL. Replaced value for buffer length arg with 0 (was (unit32_t)-1).

Method will now reuse existing buffer if reallocation fails or
allocation size is >= 2^32. Error message will be truncated and null
terminated if too long to fit in current buffer.

Additional related fixes to use uintptr_t consistently for buffer sizes
and to fix inconsistent use of str_vprintf() args and return value.

Unexpected problem found in Port.cpp. OSX needs sys/statfs.h. I don't
know how it ever got into master without including it. 

Signed-off-by: Kim Briggs <briggs@ca.ibm.com>

### CODE REVIEW

File: port/common/omrfile.c [hunk 1]
```diff
@@ -630,7 +630,7 @@ omrfile_vprintf(struct OMRPortLibrary *portLibrary, intptr_t fd, const char *for
 	COPY_VA_LIST(copyOfArgs, args);
 
 	/* What is size of buffer required ? Does not include the \0 */
```
Comment (charliegracie) [inline port/common/omrfile.c]: I thought the API did return the size including the `\0`?  Can you verify whether or not it does?  Thanks

File: port/common/j9nls.c [hunk 2]
```diff
@@ -234,34 +229,33 @@ writeSyslog(struct OMRPortLibrary *portLibrary, uintptr_t flags, const char *for
 		return;
 	}
 
-	/* Attempt to write string to on-stack buffer */
-	COPY_VA_LIST(copyOfArgs, args);
-	numberWritten = portLibrary->str_vprintf(portLibrary, outputBuffer, sizeof(outputBuffer), format, copyOfArgs);
+	char localBuffer[256];
+	char *writeBuffer = NULL;
```
Comment (charliegracie) [inline port/common/j9nls.c]: This is not valid C code.  "For C89, you must declare all of your variables at the beginning of a scope block"

File: tools/tracegen/Port.cpp [hunk 3]
```diff
@@ -30,8 +30,10 @@
 #define J9FILE_UNC_EXTENDED_LENGTH_PREFIX (L"\\\\?\\")
 #else /* defined(WIN32) */
 #include <sys/types.h>
-#if !defined(OMRZTPF)
+#if defined(AIXPPC)
 #include <sys/statvfs.h>
+#elif (defined(LINUX) && !defined(J9ZTPF)) || defined(OSX)
```
Comment (charliegracie) [inline tools/tracegen/Port.cpp]: Why are you making this change?  All of the builds were passing.  Did the compiler recently change on the travis builds? Other builds than ran today passed on OSX.

File: port/common/omrfile.c [hunk 4]
```diff
@@ -607,41 +607,33 @@ EsTranslateOpenFlags(int32_t flags)
 void
 omrfile_vprintf(struct OMRPortLibrary *portLibrary, intptr_t fd, const char *format, va_list args)
 {
-	char outputBuffer[256];
-	char *allocatedBuffer;
-	uint32_t numberWritten;
-	va_list copyOfArgs;
-
-	/* Attempt to write output to stack buffer */
-	COPY_VA_LIST(copyOfArgs, args);
```
Comment (charliegracie) [inline port/common/omrfile.c]: Why did you remove the copy of the args?

File: tools/tracegen/Port.cpp (hunk 3 above)
Comment (ktbriggs) [inline tools/tracegen/Port.cpp]: OSX needs sys/statfs.h. Compile failed for Travis build on my fork, but passed on eclipse/omr validation build for same push. Scratching head, but code inspection clearly shows problems with statvfs and statfs usage vis-a-vis includes and build flags. I believe my changes are correct.

Only explanation I can think of is that OSX builds for eclipse/omr validation include sys/statfs.h indirectly, but for briggrs/omr they do not.

File: port/common/omrfile.c (hunk 4 above)
Comment (ktbriggs) [inline port/common/omrfile.c]: I don't believe it is required. I can't find implementation for str_vprintf(), but I think it handles va_list.

File: tools/tracegen/Port.cpp (hunk 3 above)
Comment (charliegracie) [inline tools/tracegen/Port.cpp]: That code has not really changed since the initial commit.  The zptf changes recently would not have changed anything for OSX.  I am wary of changes we do not fully understand

File: port/common/j9nls.c (hunk 2 above)
Comment (ktbriggs) [inline port/common/j9nls.c]: Fixed in most recent commit.

File: tools/tracegen/Port.cpp (hunk 3 above)
Comment (ktbriggs) [inline tools/tracegen/Port.cpp]: Port.cpp changes backed out in most recent commit.

File: port/win32_include/omrportptb.h [hunk 5]
```diff
@@ -39,6 +39,7 @@
 #include "omrport.h"
 
 #define J9ERROR_DEFAULT_BUFFER_SIZE 256 /**< default customized error message size if we need to create one */
+#define J9ERROR_MAXIMUM_BUFFER_SIZE 0xFFFFFFFF /**< maximum customized error message size if we need to create one */
```
Comment (charliegracie) [inline port/win32_include/omrportptb.h]: Is this maximum buffer size correct for both 32 and 64 bit platforms? Doesn't the stdint library provide a max for unintptr_t?

File: port/unix_include/omrportptb.h [hunk 6]
```diff
@@ -41,6 +41,7 @@
 #include "omriconvhelpers.h"
 
 #define J9ERROR_DEFAULT_BUFFER_SIZE 256 /**< default customized error message size if we need to create one */
+#define J9ERROR_MAXIMUM_BUFFER_SIZE 0xFFFFFFFF /**< maximum customized error message size if we need to create one */
```
Comment (charliegracie) [inline port/unix_include/omrportptb.h]: Is this maximum buffer size correct for both 32 and 64 bit platforms? Doesn't the stdint library provide a max for unintptr_t (UINTPTR_MAX)?

File: port/include/omrportptb.h [hunk 7]
```diff
@@ -39,6 +39,7 @@
 #include "omrport.h"
 
 #define J9ERROR_DEFAULT_BUFFER_SIZE 256 /**< default customized error message size if we need to create one */
+#define J9ERROR_MAXIMUM_BUFFER_SIZE 0xFFFFFFFF /**< maximum customized error message size if we need to create one */
```
Comment (charliegracie) [inline port/include/omrportptb.h]: Is this maximum buffer size correct for both 32 and 64 bit platforms? Doesn't the stdint library provide a max for unintptr_t (UINTPTR_MAX)?

File: port/common/omrfile.c (hunk 4 above)
Comment (charliegracie) [inline port/common/omrfile.c]: You are correct it takes a va_list but as you consume args from a va_list it moves the list along and you can not get back to the beginning again.  If there are multiple calls using the same va_list I believe you will not see any values in the second, third, etc usage of the same va_list.  I believe that is why the tests were failing originally with your changes.

Comment (charliegracie) [inline port/common/omrfile.c]: You are correct it takes a va_list but as you consume args from a va_list it moves the list along and you can not get back to the beginning again. If there are multiple calls using the same va_list I believe you will not see any values in the second, third, etc usage of the same va_list. I believe that is why the tests were failing originally with your changes.

File: port/common/j9nls.c [hunk 8]
```diff
@@ -234,34 +232,30 @@ writeSyslog(struct OMRPortLibrary *portLibrary, uintptr_t flags, const char *for
 		return;
 	}
 
-	/* Attempt to write string to on-stack buffer */
-	COPY_VA_LIST(copyOfArgs, args);
-	numberWritten = portLibrary->str_vprintf(portLibrary, outputBuffer, sizeof(outputBuffer), format, copyOfArgs);
+	/* What is size of buffer required ? str_vprintf(..,NULL,..) result includes the null terminator */
+	uintptr_t bufferSize = portLibrary->str_vprintf(portLibrary, NULL, 0, format, args);
```
Comment (charliegracie) [inline port/common/j9nls.c]: You can not declare the variable here in c89 which is used on some of our platforms

File: port/common/j9nls.c [hunk 9]
```diff
@@ -234,34 +232,30 @@ writeSyslog(struct OMRPortLibrary *portLibrary, uintptr_t flags, const char *for
 		return;
 	}
 
-	/* Attempt to write string to on-stack buffer */
-	COPY_VA_LIST(copyOfArgs, args);
```
Comment (charliegracie) [inline port/common/j9nls.c]: You will need to bring back all of the COPY_VA_LIST code.  See other comments about this.

File: port/common/j9nls.c [hunk 10]
```diff
@@ -234,34 +232,30 @@ writeSyslog(struct OMRPortLibrary *portLibrary, uintptr_t flags, const char *for
...
-		return;
+	/* use local buffer if possible, allocate a buffer from system memory if local buffer not large enough */
+	if (sizeof(localBuffer) >= bufferSize) {
+		writeBuffer = localBuffer;
+	} else {
+		writeBuffer = portLibrary->mem_allocate_memory(portLibrary, bufferSize, OMR_GET_CALLSITE(), OMRMEM_CATEGORY_PORT_LIBRARY);
 	}
 
-	/* Didn't fit into outputBuffer - find out how long the buffer needs to be */
-	COPY_VA_LIST(copyOfArgs, args);
```
Comment (charliegracie) [inline port/common/j9nls.c]: You will need to bring back all of the COPY_VA_LIST code.  See other comments about this.

File: port/common/omrfile.c [hunk 11]
```diff
@@ -607,41 +607,33 @@ EsTranslateOpenFlags(int32_t flags)
...
-	/* str_vprintf always null terminates, returns number characters written excluding the null terminator */
-	if (sizeof(outputBuffer) > (numberWritten + 1)) {
-		/* write out the buffer */
-		portLibrary->file_write_text(portLibrary, fd, outputBuffer, numberWritten);
-		return;
-	}
+	char localBuffer[256];
+	char *writeBuffer = NULL;
 
-	/* Either the buffer was too small, or it was the exact size.  Unfortunately can't tell the difference,
```
Comment (charliegracie) [inline port/common/omrfile.c]: You will need to bring back all of the COPY_VA_LIST code.  See other comments about this.

File: port/common/omrfile.c [hunk 12]
```diff
@@ -607,41 +607,33 @@ EsTranslateOpenFlags(int32_t flags)
...
-	 * need to determine the size of the buffer (another call to str_vprintf) then print to the buffer,
-	 * a third call to str_vprintf
-	 */
-	COPY_VA_LIST(copyOfArgs, args);
+	/* What is size of buffer required ? str_vprintf(..,NULL,..) result includes the null terminator */
+	uintptr_t bufferSize = portLibrary->str_vprintf(portLibrary, NULL, 0, format, args);
```
Comment (charliegracie) [inline port/common/omrfile.c]: When the COPY_VA_LIST is brought back you will need to move the declaration of this variable up.

File: port/win32_include/omrportptb.h (hunk 5 above)
Comment (ktbriggs) [inline port/win32_include/omrportptb.h]: This guard is meant to catch wildly invalid results from str_vprintf(.., NULL, ...) calls on 64-bit platforms. It limits range to 32-bits (unsigned), so works for both32- and 64-bit platforms. Even 2^32 is too large to used as a reasonable limit but this will catch negative values etc. It is harmless to leave in although it would never trap an error unless str_vprintf() was badly broken.

File: port/unix_include/omrportptb.h (hunk 6 above)
Comment (ktbriggs) [inline port/unix_include/omrportptb.h]: As above.

File: port/include/omrportptb.h (hunk 7 above)
Comment (ktbriggs) [inline port/include/omrportptb.h]: As above.

File: port/common/j9nls.c (hunk 10 above)
Comment (ktbriggs) [inline port/common/j9nls.c]: All changes requested pushed in latest commit (also, copyright updates to touched files).
